new-log-viewer/src/services/LogFileManager/utils.ts

```diff
@@ -11,60 +18,56 @@ import {getBasenameFromUrlOrDefault} from "../../utils/url";
  *
- * @param beginLogEventIdx
+ * @param pageNum
+ * @param eventPositionOnPage
  * @param numEvents
  * @param pageSize
- * @return An array:
- * - pageBeginLogEventNum
- * - pageEndLogEventNum
+ * @return Matching page and log event for the cursor (i.e. cursor data.)
  */
-const getPageBoundaries = (
-    beginLogEventIdx: number,
+const getPageNumCursorData = (
+    pageNum: number,
+    eventPositionOnPage: EVENT_POSITION_ON_PAGE,
     numEvents: number,
-    pageSize: number
-): [number, number] => {
-    const pageBeginLogEventNum: number = beginLogEventIdx + 1;
+    pageSize: number,
+): CursorData => {
+    const pageBeginIdx = (pageNum - 1) * pageSize;
+    const pageEndIdx = Math.min(numEvents, pageBeginIdx + pageSize);
 
-    // Clamp ending index using total number of events.
-    const pageEndLogEventNum: number = Math.min(numEvents + 1, pageBeginLogEventNum + pageSize);
+    const matchingIdx = eventPositionOnPage === EVENT_POSITION_ON_PAGE.TOP ?
+        pageBeginIdx :
+        pageEndIdx - 1;
 
-    return [
-        pageBeginLogEventNum,
-        pageEndLogEventNum,
-    ];
+    return {pageBeginIdx, pageEndIdx, matchingIdx};
 };
 
 /**
- * Gets the data for the `PAGE_NUM` cursor.
+ * Converts a potentially "invalid" `logEventIdx` into a valid log event index. `logEventIdx` may
+ * be "invalid" if:
+ * - `logEventIdx >= numEvents`.
+ * - `logEventIdx` excluded by the current filter.
  *
- * @param pageNum
- * @param eventPositionOnPage
+ * @param logEventIdx
  * @param numEvents
- * @param pageSize
- * @return Log event numbers for:
- * - the range [begin, end) of page `pageNum`.
- * - the log event indicated by `eventPositionOnPage`.
+ * @param filteredLogEventMap
+ * @return Valid index.
  */
-const getPageNumCursorData = (
-    pageNum: number,
-    eventPositionOnPage: EVENT_POSITION_ON_PAGE,
+const getValidLogEventIdx = (
+    logEventIdx: number,
     numEvents: number,
-    pageSize: number
-): { pageBeginLogEventNum: number; pageEndLogEventNum: number; matchingLogEventNum: number } => {
-    const beginLogEventIdx = (pageNum - 1) * pageSize;
-    const [pageBeginLogEventNum, pageEndLogEventNum] = getPageBoundaries(
-        beginLogEventIdx,
-        numEvents,
-        pageSize
-    );
-    const matchingLogEventNum = eventPositionOnPage === EVENT_POSITION_ON_PAGE.TOP ?
-        pageBeginLogEventNum :
-        pageEndLogEventNum - 1;
+    filteredLogEventMap: FilteredLogEventMap,
+): number => {
+    if (null === filteredLogEventMap) {
+        // There is no filter applied.
+        return clamp(logEventIdx, 1, numEvents - 1);
+    }
+    const clampedLogEventIdx = clampWithinBounds(filteredLogEventMap, logEventIdx);
 
-    return {pageBeginLogEventNum, pageEndLogEventNum, matchingLogEventNum};
+    // Explicit cast since typescript thinks result can be null, but it can't since
+    // filteredLogEventMap has a length >= 1 and the input is clamped within the bounds
+    // of the array.
```

Comment:
```suggestion
    // Explicit cast since TypeScript thinks the return value can be null, but it can't be since
    // filteredLogEventMap is non-empty and `clampedLogEventIdx` is within the bounds of
    // `filteredLogEventMap`.
```